src/main/java/org/opensearch/security/ssl/OpenSearchSecureSettingsFactory.java

```diff
@@ -11,63 +11,124 @@
+                return Optional.of(new TransportExceptionHandler() {
                     @Override
                     public void onError(Throwable t) {
-                        sslExceptionHandler.logError(t, true);
+                        sslExceptionHandler.logError(t, false);
                     }
                 });
             }
 
             @Override
-            public Optional<ServerExceptionHandler> buildServerTransportExceptionHandler(Settings settings, TcpTransport transport) {
-                return Optional.of(new ServerExceptionHandler() {
-                    @Override
-                    public void onError(Throwable t) {
-                        sslExceptionHandler.logError(t, false);
-                    }
-                });
+            public Optional<SSLEngine> buildSecureServerTransportEngine(Settings settings, Transport transport) throws SSLException {
+                return Optional.of(sks.createServerTransportSSLEngine());
             }
 
             @Override
-            public Optional<SSLEngine> buildSecureHttpServerEngine(Settings settings, HttpServerTransport transport) throws SSLException {
-                return Optional.of(sks.createHTTPSSLEngine());
+            public Optional<SSLEngine> buildSecureClientTransportEngine(Settings settings, String hostname, int port) throws SSLException {
+                return Optional.of(sks.createClientTransportSSLEngine(hostname, port));
             }
+        });
+    }
 
+    @Override
+    public Optional<SecureHttpTransportSettingsProvider> getSecureHttpTransportSettingsProvider(Settings settings) {
+        return Optional.of(new SecureHttpTransportSettingsProvider() {
             @Override
-            public Optional<SSLEngine> buildSecureServerTransportEngine(Settings settings, TcpTransport transport) throws SSLException {
-                return Optional.of(sks.createServerTransportSSLEngine());
+            public Collection<TransportAdapterProvider<HttpServerTransport>> getHttpTransportAdapterProviders(Settings settings) {
+                return List.of(new TransportAdapterProvider<HttpServerTransport>() {
+                    @Override
+                    public String name() {
+                        return SecureNetty4HttpServerTransport.REQUEST_DECOMPRESSOR;
+                    }
+
+                    @SuppressWarnings("unchecked")
+                    @Override
+                    public <C> Optional<C> create(Settings settings, HttpServerTransport transport, Class<C> adapterClass) {
+                        if (transport instanceof SecureNetty4HttpServerTransport
+                            && ChannelInboundHandlerAdapter.class.isAssignableFrom(adapterClass)) {
+                            return Optional.of((C) new Netty4ConditionalDecompressor());
+                        } else {
+                            return Optional.empty();
```

Comment:
What do you think of throw an exception if we don't have a matching transport instance?

Thinking allowed about the possible states; If ssl disabled off, I don't believe this will be called. If we are in ssl only mode, I could go either way on checking the transport functionality, but would error on the side of throwing an exception, and in full feature ssl we only support Netty based transports.

Reply:
> What do you think of throw an exception if we don't have a matching transport instance?

So this method (or better to say provider) will be called for any supported transport: Netty4, Reactor-Netty4, Nio. At the moment, only Netty4 is supported, and for others - we do nothing (`Optional.empty()`), we just don't have this implemented. So the next provider could be picked. 

> Thinking allowed about the possible states; If ssl disabled off, I don't believe this will be called

Correct

Reply:
Just looked back into the design in OpenSearch.  I suppose that makes sense, its just decoupled a little bit more than I would expect.

As the security plugin I would want to be sure that the secure transport system has the correct relationship between REQUEST_HEADER_VERIFIER and REQUEST_DECOMPRESSOR in place.

Reply:
> As the security plugin I would want to be sure that the secure transport system has the correct relationship between REQUEST_HEADER_VERIFIER and REQUEST_DECOMPRESSOR in place.

I added test for it in previous round (in response to @scrawfor99 comment). On a general note - if we could somehow formalize that over other transport, we could end up with better solution (I just have not looked into that yet).